test/pylib/coverage_utils.py

```diff
@@ -0,0 +1,1293 @@
+
+    @staticmethod
+    @traced_func
+    async def is_debug_only(f : PathLike) -> bool :
+        """ Checks if the elf file is debug only (no code) by poking into
+        the sections in the elf
+
+        Args:
+            f (PathLike): The file to check for profile
+
+        Returns:
+            bool: True if the file is profiled, False otherwise
+        """
+        proc = await create_subprocess_shell(IS_DEBUG_ONLY_COMMAND.format(f),stdout = DEVNULL, stderr = DEVNULL)
+        await proc.wait()
+        return (proc.returncode == 0)
+    @staticmethod
+    async def get_file_type(f : PathLike) -> Self:
+        """Tries to classify the type of a file (one of FileType).
+
+        Args:
+            f (PathLike): The path to the file to be classified
+
+        Returns:
+            FileType: One of filetype, if the file couldn't be recognized,
+            FileType.UNRECOGNIZED will be returned.
+        """
+        file_description = await FileType.get_file_description(f)
+        for ft in FileType:
+            if m := ft.value[0].fullmatch(file_description):
+                if ft in ELF_TYPES:
+                    profiled =  "PROFILED_" if await FileType.is_profiled(f) else ""
+                    debug_only = "DEBUG_ONLY_" if await FileType.is_debug_only(f) else ""
+                    return FileType[debug_only + profiled + ft.name]
+                return ft
+        return FileType.UNRECOGNIZED
+
+
+ELF_TYPES = [FileType.EXEC_BIN,FileType.EXEC_SO]
+PROFILED_ELF_TYPES = [FileType.PROFILED_EXEC_BIN, FileType.PROFILED_EXEC_SO]
+
+# Convenience functions
+
+def convert_to_path_list(paths : PathLikeOrPathLikeCollection) -> List[Path]:
+    """ A convenience function to normalize PathLikeOrPathLikeCollection into
+        a PathCollection
+
+    Args:
+        paths (PathLikeOrPathLikeCollection): The parameter to convert
+
+    Returns:
+        List[Path]: A path list
+    """
+    if paths is None:
+        return []
+    elif isinstance(paths, str) or not isinstance(paths, IterableType):
+        return [Path(paths)]
+    return [ Path(path) for path in paths]
+
+def concurrency_to_semaphore(concurrency : ConcurrencyParam) -> asyncio.Semaphore:
+    """A convenience function to convert a `ConcurrencyParam` to be used to limit
+    concurrency of execution.
+
+    Args:
+        concurrency (ConcurrencyParam): The concurrency to be converted to a semaphore.
+
+    Returns:
+        asyncio.Semaphore: A semaphore according to the following semantics:
+        1. If concurrence's actual type is int, so a semaphore with `concurrency` units.
+        2. If `concurrency` is a semaphore, it will be returned as is
+        3. In any other case, a semaphore with 1 unit will be returned (no concurrency allow)
+    """
+    if isinstance(concurrency,int):
+        return asyncio.Semaphore(concurrency)
+    elif isinstance(concurrency,asyncio.Semaphore):
+        return concurrency
+    else:
+        return asyncio.Semaphore(1)
```

Comment:
I don't understand why this is needed, doesn't python already do this?

Don't we use just one concurrency library? Can't we write asyncio.Semaphore(concurrency) at the call site instead of this function?

Reply:
This function is to allow an external semaphore to be given to the library, imagine a case where you want a
function to share the concurrency with other functions in your program.
It is not mandatory but allows greater flexibility when used as a library.